src/sage/combinat/sf/sfa.py

```diff
@@ -6379,6 +6424,253 @@ def exponential_specialization(self, t=None, q=1):
+            sage: F, R = P.construction()
+            sage: F(ZZ["t"].hom(GF(3)["t"]))
+            Generic morphism:
+              From: Symmetric Functions over Univariate Polynomial Ring in t over Integer Ring in the Jack P basis
+              To:   Symmetric Functions over Univariate Polynomial Ring in t over Finite Field of size 3 in the Jack P basis
+
+            sage: R.<q, t> = ZZ[]
+            sage: H = SymmetricFunctions(R).macdonald().H()
+            sage: F, R = H.construction()
+            sage: F(ZZ["q", "t"].hom(GF(3)["q", "t"]))  # known bug
+        """
+        dom = self(f.domain())
+        codom = self(f.codomain())
+
+        def action(x):
+            return codom._from_dict({a: f(b)
+                                     for a, b in x.monomial_coefficients().items()})
+        return dom.module_morphism(function=action, codomain=codom)
+
+    def __eq__(self, other):
+        """
+        EXAMPLES::
+
+            sage: Sym = SymmetricFunctions(QQ)
+            sage: qbar1 = Sym.qbar(q=1/2)
+            sage: qbar2 = Sym.qbar(q=1)
+            sage: qbar1.construction()[0] == qbar2.construction()[0]
+            False
+        """
+        if not isinstance(other, type(self)):
+            return False
+        return (self._basis == other._basis
+                and self._name == other._name
+                and self._args == other._args)
+
+    def __hash__(self):
+        """
+        Return the hash of ``self``.
+
+        EXAMPLES::
+
+            sage: Sym = SymmetricFunctions(QQ)
+            sage: F1 = Sym.qbar(q=1/2).construction()[0]
+            sage: F2 = Sym.qbar(q=1).construction()[0]
+            sage: hash(F1) == hash(F2)
+            False
+        """
+        return hash(repr(self))
+
+    def _repr_(self):
+        """
+        Return a string representation of ``self``.
+
+        EXAMPLES::
+
+            sage: w = SymmetricFunctions(ZZ).witt()
+            sage: F, R = w.construction()
+            sage: F
+            SymmetricFunctionsFunctor[Witt]
+
+            sage: R.<q, t> = ZZ[]
+            sage: H = SymmetricFunctions(R).macdonald().H()
+            sage: F, R = H.construction()
+            sage: F
+            SymmetricFunctionsFunctor[Macdonald H]
+        """
+        return "SymmetricFunctionsFunctor[" + self._name + "]"
+
+
+class SymmetricFunctionsFamilyFunctor(SymmetricFunctionsFunctor):
+    def __init__(self, basis, family, name, *args):
+        r"""
+        Initialize the functor.
+
+        INPUT:
+
+        - ``basis`` -- the basis of the symmetric function algebra
+
+        .. WARNING:
+
+            Strictly speaking, this is not necessarily a functor on
+            :class:`CommutativeRings`, but rather a functor on
+            commutative rings with some distinguished elements.  For
+            example, for the Macdonald polynomials, we have to
+            specify `q` and `t` in the ring.  Apart from that, the
```

Comment:
Shouldn't the functor just store the names of these elements, as is done in the existing construction functors `PolynomialFunctor` etc.?
Right now, it's remembering their parent, i.e., the whole `Sym.base_ring()` which seems too much

Reply:
But let's say specifically for `Macdonald` -- won't storing `q` and `t` as elements bake something into the construction functor that should be determined by the base ring that it is applied to?

Reply:
This absolutely should *not* store the names because they are not names! They are elements in the ground ring. You cannot guarantee that a string representation of these elements will be convertible back into the base ring (even less so if you change the ring). For example, if we create symmetric functions over symmetric functions:
```
sage: s = SymmetricFunctions(QQ).s()
sage: ss = SymmetricFunctions(s).s()
sage: ss[2]
s[]*s[2]
sage: SymmetricFunctions(s).macdonald(s[2], s[3])
Macdonald polynomials with q=s[2] and t=s[3] over Symmetric Functions over Rational Field in the Schur basis
```
As indicated by the note, the Macdonald case (for example) really should be a functor from "doubly pointed" commutative rings to commutative rings. Making this mathematically precise is a bit cumbersome, but for the practical purposes at hand, this is the correct way to view things.